lua/packer/snapshot.lua

```diff
@@ -0,0 +1,192 @@
+  end
+
+  local dir = vim.loop.fs_opendir(config.snapshot_path)
+
+  if dir ~= nil then
+    local res = vim.loop.fs_readdir(dir)
+    while res ~= nil do
+      for _, entry in ipairs(res) do
+        if entry.type == "file" and vim.startswith(entry.name, lead) then
+          completion_list[#completion_list + 1] = entry.name
+        end
+      end
+
+      res = vim.loop.fs_readdir(dir)
+    end
+  end
+
+  vim.loop.fs_closedir(dir)
+  return completion_list
+end
+
+--- Completion for listing single plugins before taking snapshot
+--- Intended to provide completion for PackerSnapshot command
+snapshot.completion.create = function (lead, cmdline, pos)
+  local cmd_args = (vim.fn.split(cmdline, " "))
+
+  if #cmd_args > 1 then
+    return plugin_complete(lead, cmdline, pos)
+  end
+
+  return {}
+end
+
+--- Completion for listing snapshots in `config.snapshot_path` and single plugins after
+--- the first argument is provided
+--- Intended to provide completion for PackerRollback command
+snapshot.completion.rollback = function (lead, cmdline, pos)
+  local cmd_args = vim.split(cmdline, " ")
+
+  if #cmd_args > 2 then
+    return plugin_complete(lead)
+  else
+    return snapshot.completion.snapshot(lead, cmdline, pos)
+  end
+end
+
+---Serializes a table of git-plugins with `short_name` as table key and another
+---table with `commit`; the serialized tables will be written in the path `snapshot_path`
+---provided, if there is already a snapshot it will be overwritten
+---Snapshotting work only with `plugin_utils.git_plugin_type` type of plugins,
+---other will be ignored.
+---@param snapshot_path string realpath for snapshot file
+---@param plugins table<string, any>[]
+snapshot.create = function(snapshot_path, plugins)
+  assert(type(snapshot_path) == "string",
+    fmt("filename needs to be a string but '%s' provided", type(snapshot_path)))
+  assert(type(plugins) == "table",
+    fmt("plugins needs to be an array but '%s' provided", type(plugins)))
+
+  return async(function()
+    local snapshot_plugins = {}
+    local installed = {}
+    local opt, start = plugin_utils.list_installed_plugins()
+
+    for key, _ in pairs(opt) do installed[key] = key end
+    for key, _ in pairs(start) do installed[key] = key end
+
+    for _, plugin in pairs(plugins) do
+      if installed[plugin.install_path] then -- this plugin is installed
+        log.debug(fmt("Snapshotting '%s'", plugin.short_name))
+        if plugin.type == plugin_utils.git_plugin_type then
+          local rev = await(plugin.get_rev())
+
+          if rev.err then
+            local warn = fmt("Snapshotting %s failed because of error '%s'",
+              plugin.short_name,
+              rev.err
+            )
+            log.warn(warn)
+          else
+            snapshot_plugins[plugin.short_name] = {commit = rev.ok}
+          end
+        end
+      end
+    end
+
+    ---@type string
+    local snapshot_content = "return " .. vim.inspect(snapshot_plugins)
```

Comment:
Since it's just data . Maybe we can consider storing it serialized as json/msgpack instead of putting it as a lua file and evaluating it . What do you think ?

Reply:
My first thought was the same, the problem is that I am not aware of any kind of way to deserialize JSON in Lua without implementing a parser on your own.

Reply:
There's `vim.mpack` (msgpack implementation) & `vim.json` (cjson implementation) . `vim.json.encode` & `vim.json.decode` is better then `json_encode` and `json_decode` but I think this was added on nvim-0.5.1 / 0.6.0 I'm not sure which . @wbthomason what your opinion on using 0.5+ features ?

Reply:
> There's `vim.mpack` (msgpack implementation) & `vim.json` (cjson implementation) . `vim.json.encode` & `vim.json.decode` is better then `json_encode` and `json_decode` but I think this was added on nvim-0.5.1 / 0.6.0 I'm not sure which . @wbthomason what your opinion on using 0.5+ features ?

I could just wrap it in util.
If the user is running 0.5 then use json_encode, else if 0.6+ use vim.json.encode etc.

Reply:
> I could just wrap it in utils.
If the user is running 0.5 then use json_encode, else if 0.6+ use vim.json.encode etc.

Now thinking about it, for such small files difference in using vim.json / json_encode|decode won't be significant . Using different functions based on nvim version is unnecessary complexity . We'll be fine with just using json_encode|decode.